drivers/partition/partition.c

```diff
@@ -0,0 +1,255 @@
+	if (result != 0) {
+		WARN("Failed to seek (%i)\n", result);
+		return result;
+	}
+	result = io_read(image_handle, (uintptr_t)&mbr_sector,
+			 PARTITION_BLOCK_SIZE, &bytes_read);
+	if (result != 0) {
+		WARN("Failed to read data (%i)\n", result);
+		return result;
+	}
+
+	/* Check MBR boot signature. */
+	if ((mbr_sector[PARTITION_BLOCK_SIZE - 2] != 0x55) ||
+	    (mbr_sector[PARTITION_BLOCK_SIZE - 1] != 0xAA)) {
+		return -EINVAL;
+	}
+	/* Check the first partition entry. */
+	entry = (mbr_entry_t *)((uintptr_t)&mbr_sector +
+				MBR_PRIMARY_ENTRY_OFFSET);
+	if (entry->type == PARTITION_TYPE_GPT)
+		return -EAGAIN;
+	return -EINVAL;
+}
+
+static void dump_entries(int num)
+{
+	char name[EFI_NAMELEN];
+	int i, j, len;
+
+	INFO("Partition table with %d entries:\n", num);
+	for (i = 0; i < num; i++) {
+		len = snprintf(name, EFI_NAMELEN, "%s", entry_list[i].name);
+		for (j = 0; j < EFI_NAMELEN - len; j++) {
+			name[len + j] = ' ';
+		}
+		INFO("%d: %s %lx-%lx\n", i + 1, name, entry_list[i].start,
+		     entry_list[i].start + entry_list[i].length - 4);
+	}
+}
+
+static int verify_partition_gpt(uintptr_t image_handle)
+{
+	gpt_header_t header;
+	gpt_entry_t entry;
+	size_t bytes_read;
+	int result, i;
+
+	result = io_seek(image_handle, IO_SEEK_SET, GPT_HEADER_OFFSET);
+	if (result != 0) {
+		return result;
+	}
+	result = io_read(image_handle, (uintptr_t)&header,
+			 sizeof(gpt_header_t), &bytes_read);
+	if ((result != 0) || (sizeof(gpt_header_t) != bytes_read)) {
+		return result;
+	}
+	if (memcmp(header.signature, GPT_SIGNATURE,
+		   sizeof(header.signature)) != 0) {
+		return -EINVAL;
+	}
+	part_num_entries = header.part_num;
+	if (part_num_entries > MAX_PARTITION_NUM) {
+		return -EINVAL;
+	}
+
+	result = io_seek(image_handle, IO_SEEK_SET, GPT_ENTRY_OFFSET);
+	if (result != 0) {
+		return result;
+	}
+	for (i = 0; i < part_num_entries; i++) {
+		result = io_read(image_handle, (uintptr_t)&entry,
+				 sizeof(gpt_entry_t), &bytes_read);
+		if (result || (sizeof(gpt_entry_t) != bytes_read)) {
+			return result;
+		}
+		result = parse_gpt_entry(&entry, i);
+		if (result != 0) {
+			break;
+		}
+	}
+	if (i == 0) {
+		return -EINVAL;
+	}
+	part_num_entries = i;
+	dump_entries(part_num_entries);
+
+	return 0;
+}
+
+int load_partition_table(unsigned int image_id)
+{
+	uintptr_t dev_handle, image_handle, image_spec = 0;
+	int result;
+
+	result = plat_get_image_source(image_id, &dev_handle, &image_spec);
+	if (result != 0) {
+		WARN("Failed to obtain reference to image id=%u (%i)\n",
+			image_id, result);
+		return result;
+	}
+
+	result = io_open(dev_handle, image_spec, &image_handle);
+	if (result != 0) {
+		WARN("Failed to access image id=%u (%i)\n", image_id, result);
+		return result;
+	}
+
+	result = verify_partition_mbr(image_handle);
+	if (result != 0) {
+		INFO("Can't find MBR partition table (%i)\n", result);
+		if (result != -EAGAIN)
+			goto exit;
+	} else {
+		/* Find MBR partition table. */
+		goto exit;
+	}
+
```

Comment:
Need a comment to say that MBR is parsed OK and now need to verify that first entry is a GPT entry.